move -DDEBUG for ASan builds in memory.c (closes #5013)

AlekseyCherepanov opened the pull request:
BTW there is such check in `memory.c`:
```c
#if defined(WITH_ASAN) || defined(DEBUG)
```

It could be reduced to just `defined(DEBUG)` because `WITH_ASAN` implies `DEBUG` now.

AlekseyCherepanov committed:
Moved -DDEBUG for memory.c in builds with ASan after inclusion of misc.h

(closes #5013)

AlekseyCherepanov committed:
includes of listconf.h: changed comment to mention misc.h as requirement

AlekseyCherepanov commented:
`john.c` includes `listconf.h` and does not have comment as in `options.c` and `listconf.c`. If it is not needed to reduce difference from john proper, then I would add the comment.

solardiz commented:
This is working so well that we got ASan failures from CI right on this PR. Do we merge it first, or do we get those bugs fixed first?

solardiz commented:
`non-SIMD`:

```
Testing: krb5-18, Kerberos 5 DB etype 18 [DES / PBKDF2-SHA1 32/64 AES]... (2xOMP) =================================================================
==5615==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60300013016f at pc 0x00000071f5b3 bp 0x7ffd626004f0 sp 0x7ffd626004e0
WRITE of size 1 at 0x60300013016f thread T0
    #0 0x71f5b2 in get_binary /home/JtR/project/src/krb5_db_fmt_plug.c:260
    #1 0x86c8bd in ldr_load_pw_line /home/JtR/project/src/loader.c:985
    #2 0x8700bf in ldr_init_test_db /home/JtR/project/src/loader.c:1317
    #3 0x830590 in benchmark_all /home/JtR/project/src/bench.c:872
    #4 0x861122 in john_run /home/JtR/project/src/john.c:1661
    #5 0x861122 in main /home/JtR/project/src/john.c:2082
    #6 0x7f7014ec8041 in __libc_start_main (/lib64/libc.so.6+0x27041)
    #7 0x4061bd in _start (/home/JtR/project/run/john+0x4061bd)

0x60300013016f is located 0 bytes to the right of 31-byte region [0x603000130150,0x60300013016f)
allocated by thread T0 here:
    #0 0x7f7015742667 in __interceptor_malloc (/lib64/libasan.so.6+0xb0667)
    #1 0x884d23 in mem_alloc /home/JtR/project/src/memory.c:90
    #2 0x884fd7 in mem_alloc_tiny /home/JtR/project/src/memory.c:201
    #3 0x71f48d in get_binary /home/JtR/project/src/krb5_db_fmt_plug.c:256
    #4 0x86c8bd in ldr_load_pw_line /home/JtR/project/src/loader.c:985
    #5 0x8700bf in ldr_init_test_db /home/JtR/project/src/loader.c:1317
    #6 0x830590 in benchmark_all /home/JtR/project/src/bench.c:872
    #7 0x861122 in john_run /home/JtR/project/src/john.c:1661
    #8 0x861122 in main /home/JtR/project/src/john.c:2082
    #9 0x7f7014ec8041 in __libc_start_main (/lib64/libc.so.6+0x27041)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/JtR/project/src/krb5_db_fmt_plug.c:260 in get_binary
Shadow bytes around the buggy address:
  0x0c068001dfd0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c068001dfe0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c068001dff0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c068001e000: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c068001e010: fa fa fa fa fa fa fa fa fd fd fd fa fa fa 00 00
=>0x0c068001e020: 07 fa fa fa 00 00 07 fa fa fa 00 00 00[07]fa fa
  0x0c068001e030: fd fd fd fa fa fa fd fd fd fa fa fa fd fd fd fa
  0x0c068001e040: fa fa fd fd fd fa fa fa fd fd fd fa fa fa fd fd
  0x0c068001e050: fd fa fa fa fd fd fd fa fa fa fd fd fd fa fa fa
  0x0c068001e060: fd fd fd fa fa fa fd fd fd fa fa fa fd fd fd fa
  0x0c068001e070: fa fa 00 00 07 fa fa fa 00 00 07 fa fa fa 00 00
```

The `asan` build's crash is similar.

solardiz commented:
@AlekseyCherepanov Would you push a commit fixing `krb5-18` to this PR, so we'll see it's all-green before merging?

AlekseyCherepanov commented:
Ok. I'll try to fix that before merging.

AlekseyCherepanov commented:
`--test=0 --format=krb5-18` does not crash. `--test=0 --format=krb5-17,krb5-18` crashes.

AlekseyCherepanov committed:
krb5 formats: fix overflow allocating max size suitable for all formats

AlekseyCherepanov commented:
Cool bug: there is `get_binary()` shared for several formats. It has static variable to hold pointer to buffer. Buffer is allocated with size precisely for the first format. So `krb5-17` gets size 16, then `krb5-18` reuses the buffer and overflows because it needs size 32. No problem if `krb5-18` is the first (e.g. called alone).

I fixed the size to be 32. I use explicit `32` instead of `BINARY_SIZE_18` because code above uses explicit constants.

I added a comment. Is it needed?

I thought of `assert(binary_size <= 32)` but I guess it is not necessary.

AlekseyCherepanov commented:
With fixed value, initialization of static variable could be moved to the beginning of function.
